bigquery/google/cloud/bigquery/magics.py

```diff
@@ -0,0 +1,281 @@
+
+        Example:
+            Manually setting the context credentials:
+
+            >>> from google.cloud.bigquery import magics
+            >>> from google.oauth2 import service_account
+            >>> credentials = (service_account
+            ...     .Credentials.from_service_account_file(
+            ...         '/path/to/key.json'))
+            >>> magics.context.credentials = credentials
+
+
+        .. _auth docs: http://google-auth.readthedocs.io
+            /en/latest/user-guide.html#obtaining-credentials
+        """
+        if self._credentials is None:
+            self._credentials, _ = google.auth.default()
+        return self._credentials
+
+    @credentials.setter
+    def credentials(self, value):
+        self._credentials = value
+
+    @property
+    def project(self):
+        """str: Default project to use for queries performed through IPython
+        magics
+
+        Note:
+            The project does not need to be explicitly defined if you have an
+            environment default project set. If you do not have a default
+            project set in your environment, manually assign the project as
+            demonstrated in the example below.
+
+        Example:
+            Manually setting the context project:
+
+            >>> from google.cloud.bigquery import magics
+            >>> magics.context.project = 'my-project'
+        """
+        if self._project is None:
+            _, self._project = google.auth.default()
+        return self._project
+
+    @project.setter
+    def project(self, value):
+        self._project = value
+
+
+context = Context()
+
+
+def _run_query(client, query, job_config=None):
+    """Runs a query while printing status updates
+
+    Args:
+        client (google.cloud.bigquery.client.Client):
+            Client to bundle configuration needed for API requests.
+        query (str):
+            SQL query to be executed. Defaults to the standard SQL dialect.
+            Use the ``job_config`` parameter to change dialects.
+        job_config (google.cloud.bigquery.job.QueryJobConfig, optional):
+            Extra configuration options for the job.
+
+    Returns:
+        google.cloud.bigquery.job.QueryJob: the query job created
+
+    Example:
+        >>> client = bigquery.Client()
+        >>> _run_query(client, "SELECT 17")
+        Executing query with job ID: bf633912-af2c-4780-b568-5d868058632b
+        Query executing: 1.66s
+        Query complete after 2.07s
+        'bf633912-af2c-4780-b568-5d868058632b'
+    """
+    start_time = time.time()
+    query_job = client.query(query, job_config=job_config)
+    print('Executing query with job ID: {}'.format(query_job.job_id))
+
+    while True:
+        print('\rQuery executing: {:0.2f}s'.format(
+            time.time() - start_time), end='')
+        try:
+            query_job.result(timeout=0.5)
+        except futures.TimeoutError:
+            continue
+        break
```

Comment:
this break is in a surprising spot. Would it make more sense to put it as the last statement in the `try`?